File: examples/SmartHomeDemo/src/User/SmartHome/HueBulb.cs
using System;
using System.Collections.Generic;

namespace SmartHome;

public partial class HueBulb
{
    public HueBulb(string deviceId, string displayName)
    {
        DeviceId = deviceId;
        DisplayName = displayName;
    }

    public HashSet<DeviceFeature> GetSupportedFeatures()
    {
        return
        [
            DeviceFeature.Lighting,
            DeviceFeature.Automation,
        ];
    }

    public string DescribeState()
    {
        return BrightnessPercent == 0
            ? "off"
            : $"on at {BrightnessPercent}% brightness";
    }

    public void SetBrightness(int value)
    {
        BrightnessPercent = Math.Clamp(value, 0, 100);
    }
}

File: examples/SmartHomeDemo/cpp/user/src/SmartHome/HueBulb.cpp
#include "SmartHome/HueBulb.hpp"

#include <algorithm>
#include <utility>

namespace SmartHome
{

HueBulb::HueBulb(::Crimson::Cpp::String deviceId, ::Crimson::Cpp::String displayName)
{
    SetDeviceId(std::move(deviceId));
    SetDisplayName(std::move(displayName));
}

::Crimson::Cpp::Set<DeviceFeature> HueBulb::GetSupportedFeatures()
{
    return
    {
        DeviceFeature::Lighting,
        DeviceFeature::Automation,
    };
}

::Crimson::Cpp::String HueBulb::DescribeState()
{
    return GetBrightnessPercent() == 0
        ? "off"
        : "on at " + std::to_string(GetBrightnessPercent()) + "% brightness";
}

void HueBulb::SetBrightness(std::int32_t value)
{
    SetBrightnessPercent(std::clamp<std::int32_t>(value, 0, 100));
}

}
